pattern/src/lib.rs

```diff
@@ -1,9 +1,107 @@
+//! # use prjunnamed_pattern::patterns::*;
+//! let matcher = netlist_matches! {
+//!     [PAdc@y [PAny@a] [PZero] [PInput@c ("cin")]] => (y, a, c);
+//! };
+//! # let design: Design = todo!();
+//! # let target: Value = todo!();
+//! # let result: Option<(Value, Value, Net)> = matcher(&design, &target);
+//! ```
+//!
+//! A pattern call is delimited with square brackets and consists of the
+//! following parts:
+//! - the name of the pattern being called, e.g. `PAdc`
+//! - a result capture, e.g. `@y`. This is optional, and if it is not provided,
+//!   the value captured by the pattern is discarded.
+//!
+//!   Each pattern captures exactly one value. Often, it is the [`Net`] or
+//!   [`Value`] it is being matched against. In rare cases, the captured value
+//!   is of the unit type `()`, in which case it is usually discarded by the
+//!   user.
+//! - a number of expression arguments, e.g. `("cin")`. Expression arguments
+//!   are evaluated according to normal Rust rules, and the results are passed
+//!   to the pattern. Each pattern has a fixed number of expression arguments
+//!   it expects.
+//!
+//!   If multiple expression arguments are being passed to the pattern, each
+//!   requires its own pair of parentheses, e.g. `PFeline@cat ("meow") (42)`
+//! - a number of pattern arguments, e.g. `[PZero]`. Pattern arguments
+//!   follow the pattern call syntax being described.
+//!
+//!   Each pattern has a fixed number of pattern arguments it expects.
+//!   Most commonly, they correspond to the inputs of the cell being matched.
+//!
+//!   If you do not wish to perform further matching on the inputs, use `[PAny]`.
+//!
+//! Much like `match`, [`netlist_matches!`] supports `if` and `if let` guards.
+//! For example:
+//!
+//! ```no_run
+//! # use prjunnamed_netlist::{Design, Net, Value};
+//! # use prjunnamed_pattern::netlist_matches;
+//! # use prjunnamed_pattern::patterns::*;
+//! let matcher = netlist_matches! {
+//!     [PXor [PAny@a] [PAny@b]] if a == b => Value::zero(a.len());
+//! };
+//! # let design: Design = todo!();
+//! # let target: Value = todo!();
+//! # let result: Option<Value> = matcher(&design, &target);
+//! ```
+
+/// A pattern that can be matched against a `Target`.
+///
+/// Most commonly, `Target` will be [`Value`] or [`Net`], but it could also be
+/// something like `u32`.
+///
+/// A pattern must also have a `new` function that will be called by the macro
+/// to construct an instance of the pattern. The arguments of the `new` function
+/// consist of all the expression arguments, followed by all the pattern
+/// arguments.
 pub trait Pattern<Target> {
+    /// The capture type resulting from matching this pattern *and all the
+    /// patterns passed as arguments*.
+    ///
+    /// The macros expect this to be a tuple, where the first element is
+    /// the value captured by the pattern itself, and the following elements
+    /// correspond to the `Capture`s of all the child patterns.
```

Comment:
I wouldn't say "the macros expect" specifically since what you're describing here is a contract of the library/trait itself (albeit captured only in documentation). The macros are a syntax that relies on this contract rather than adding restrictions in themselves. (This is minor.)